pkg/labels/consul_applicator.go

```diff
@@ -208,7 +210,9 @@ func convertLabeledToKVP(l Labeled) (*api.KVPair, error) {
 // to the cost of querying for this subtree on any sizeable fleet of machines. Instead, preparers should
 // use the httpApplicator from a server that exposes the results of this (or another)
 // implementation's watch.
-func (c *consulApplicator) WatchMatches(selector labels.Selector, labelType Type, quitCh chan struct{}) chan *[]Labeled {
+func (c *consulApplicator) WatchMatches(selector labels.Selector, labelType Type, quitCh chan struct{}) chan WatchResult {
+	c.aggregatorMux.Lock()
+	defer c.aggregatorMux.Unlock()
```

Comment:
I'm a little nervous about releasing this lock after calling `Wait()` given the complexity of that method. Can you release the lock before? `Wait()` can handle the extra concurrency.